// drizzle-orm/node-postgres/index.cjs
'use strict';

var pg = require('pg');
var logger = require('../logger-caa1ca6e.cjs');
var session = require('../session-08a2429e.cjs');
var relations = require('../relations-421f2ba6.cjs');
require('../errors-d0192d62.cjs');

const { Pool } = pg;
class NodePgPreparedQuery extends session.PreparedQuery {
    constructor(client, queryString, params, logger, fields, name, customResultMapper) {
        super();
        this.client = client;
        this.params = params;
        this.logger = logger;
        this.fields = fields;
        this.customResultMapper = customResultMapper;
        this.rawQuery = {
            name,
            text: queryString,
        };
        this.query = {
            name,
            text: queryString,
            rowMode: 'array',
        };
    }
    async execute(placeholderValues = {}) {
        const params = relations.fillPlaceholders(this.params, placeholderValues);
        this.logger.logQuery(this.rawQuery.text, params);
        const { fields, rawQuery, client, query, joinsNotNullableMap, customResultMapper } = this;
        if (!fields && !customResultMapper) {
            return client.query(rawQuery, params);
        }
        const result = await client.query(query, params);
        return customResultMapper
            ? customResultMapper(result.rows)
            : result.rows.map((row) => relations.mapResultRow(fields, row, joinsNotNullableMap));
    }
    all(placeholderValues = {}) {
        const params = relations.fillPlaceholders(this.params, placeholderValues);
        this.logger.logQuery(this.rawQuery.text, params);
        return this.client.query(this.rawQuery, params).then((result) => result.rows);
    }
    values(placeholderValues = {}) {
        const params = relations.fillPlaceholders(this.params, placeholderValues);
        this.logger.logQuery(this.rawQuery.text, params);
        return this.client.query(this.query, params).then((result) => result.rows);
    }
}
class NodePgSession extends session.PgSession {
    constructor(client, dialect, schema, options = {}) {
        super(dialect);
        this.client = client;
        this.schema = schema;
        this.options = options;
        this.logger = options.logger ?? new logger.NoopLogger();
    }
    prepareQuery(query, fields, name, customResultMapper) {
        return new NodePgPreparedQuery(this.client, query.sql, query.params, this.logger, fields, name, customResultMapper);
    }
    async query(query, params) {
        this.logger.logQuery(query, params);
        const result = await this.client.query({
            rowMode: 'array',
            text: query,
            values: params,
        });
        return result;
    }
    async queryObjects(query, params) {
        return this.client.query(query, params);
    }
    async transaction(transaction, config) {
        const session = this.client instanceof Pool
            ? new NodePgSession(await this.client.connect(), this.dialect, this.schema, this.options)
            : this;
        const tx = new NodePgTransaction(this.dialect, session, this.schema);
        await tx.execute(relations.sql `begin${config ? relations.sql ` ${tx.getTransactionConfigSQL(config)}` : undefined}`);
        try {
            const result = await transaction(tx);
            await tx.execute(relations.sql `commit`);
            return result;
        }
        catch (error) {
            await tx.execute(relations.sql `rollback`);
            throw error;
        }
        finally {
            if (this.client instanceof Pool) {
                session.client.release();
            }
        }
    }
}
class NodePgTransaction extends session.PgTransaction {
    async transaction(transaction) {
        const savepointName = `sp${this.nestedIndex + 1}`;
        const tx = new NodePgTransaction(this.dialect, this.session, this.schema, this.nestedIndex + 1);
        await tx.execute(relations.sql.raw(`savepoint ${savepointName}`));
        try {
            const result = await transaction(tx);
            await tx.execute(relations.sql.raw(`release savepoint ${savepointName}`));
            return result;
        }
        catch (err) {
            await tx.execute(relations.sql.raw(`rollback to savepoint ${savepointName}`));
            throw err;
        }
    }
}

const { types } = pg;
class NodePgDriver {
    constructor(client, dialect, options = {}) {
        this.client = client;
        this.dialect = dialect;
        this.options = options;
        this.initMappers();
    }
    createSession(schema) {
        return new NodePgSession(this.client, this.dialect, schema, { logger: this.options.logger });
    }
    initMappers() {
        types.setTypeParser(types.builtins.TIMESTAMPTZ, (val) => val);
        types.setTypeParser(types.builtins.TIMESTAMP, (val) => val);
        types.setTypeParser(types.builtins.DATE, (val) => val);
    }
}
function drizzle(client, config = {}) {
    const dialect = new relations.PgDialect();
    let logger$1;
    if (config.logger === true) {
        logger$1 = new logger.DefaultLogger();
    }
    else if (config.logger !== false) {
        logger$1 = config.logger;
    }
    let schema;
    if (config.schema) {
        const tablesConfig = relations.extractTablesRelationalConfig(config.schema, relations.createTableRelationsHelpers);
        schema = {
            fullSchema: config.schema,
            schema: tablesConfig.tables,
            tableNamesMap: tablesConfig.tableNamesMap,
        };
    }
    const driver = new NodePgDriver(client, dialect, { logger: logger$1 });
    const session$1 = driver.createSession(schema);
    return new session.PgDatabase(dialect, session$1, schema);
}

exports.NodePgDriver = NodePgDriver;
exports.NodePgPreparedQuery = NodePgPreparedQuery;
exports.NodePgSession = NodePgSession;
exports.NodePgTransaction = NodePgTransaction;
exports.drizzle = drizzle;
